### PR TITLE
feat(occ): Add a --debug option to output all log levels to the output

### PR DESCRIPTION
**Pull request by come-nc**
Related to #30989 

## Summary

Adds a universal option to pass to occ commands so that all levels of logs are printed in the output, even if the loglevel on the instance is higher.
This allows to inspect all the logging related to a command without having it mixed up with logs from other stuff going on in the server. Also allows to have logs and command output in the same place to see in which order things happen.
This is especially useful with user_ldap as all LDAP calls are logged in debug level.

## TODO

- [x] Decide on the correct naming for the option, currently `--debug` bug maybe this will cause conflicts too easily. **Changed to `--debug-log`**
- [ ] Decide whether to document the option in the usage output for classes based on `OC\Core\Command\Base`.
- [x] Determine whether the output format is the right one. (Do we want raw json instead?) **Abandoned this idea, there is not much data to put in json anyway**
- [x] Do we need to be able to do the same with an arbitrary loglevel, like --debug=1 to only show info and above? **Added support for `--debug-log-level` when `--debug-log` is used**
- [x] Should we activate the profiler when this option is used and output the profile? **Not automatically**
- [x] How can we make it output DB requests to the output as well? The profiler idea above would work. **Left as an exercise for later**

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by come-nc** [comment]
Hum, I think this is a duplicate of https://github.com/nextcloud/logreader/pull/873 .
I’d prefer having this in core and as an option though, as setting an env var is not always the same syntax for customers, and `logreader` may be disabled. The point was to have a command we can pass to any user to get a detailed idea of what happens when they run a command.

File: core/Listener/BeforeMessageLoggedEventListener.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+	/**
+	 * Register listener to log messages and remove debug options from $_SERVER['argv']
+	 */
+	public static function setup(): void {
+		$eventDispatcher = Server::get(IEventDispatcher::class);
+		$argv = $_SERVER['argv'];
+		$level = 0;
+		foreach ($argv as $key => $arg) {
+			if ($arg === '--debug-log') {
```
Comment (nickvergessen) [inline core/Listener/BeforeMessageLoggedEventListener.php:57]: Should we put them into OCP somewhere so app devs have a list/track of "reserved arguments"?

**Review by nickvergessen — APPROVED** [review]

File: lib/base.php [hunk 2]
```diff
@@ -578,6 +579,10 @@ public static function init(): void {
 		self::$server = new \OC\Server(\OC::$WEBROOT, self::$config);
 		self::$server->boot();
 
+		if (self::$CLI && in_array('--debug-log', $_SERVER['argv'])) {
```
Comment (artonge) [inline lib/base.php:582]: Isn't it possible?

```suggestion
		if (self::$CLI && in_array(ReservedOptions::DEBUG_LOG, $_SERVER['argv'])) {
```

**Review by artonge — APPROVED** [review]

File: lib/public/Console/ReservedOptions.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCP\Console;
+
+/**
+ * @since 30.0.0
+ */
+final class ReservedOptions {
```
Comment (nickvergessen) [inline lib/public/Console/ReservedOptions.php:15]: @come-nc Please document in https://docs.nextcloud.com/server/latest/developer_manual/app_publishing_maintenance/app_upgrade_guide/upgrade_to_30.html#id1

**Pull request merged**